### PR TITLE
Emtf dqm emulator data comparison 

### PR DESCRIPTION
**Pull request by EsmaeelEskandari**

### CODE REVIEW

**Comment by EsmaeelEskandari** [comment]
Dear @dmitrijus, I followed your recipe and made these 2 PRs: #16869 and #16870 for 90x and 81x respectively.

**Comment by dmitrijus** [comment]
+1

File: DQM/L1TMonitor/src/L1TStage2EMTF.cc [hunk 1]
```diff
@@ -169,45 +191,62 @@ void L1TStage2EMTF::bookHistograms(DQMStore::IBooker& ibooker, const edm::Run&,
...
     emtfMuonhwQual->setBinLabel(bin, std::to_string(bin - 1), 1);
   }
+
 }
 
 void L1TStage2EMTF::analyze(const edm::Event& e, const edm::EventSetup& c) {
 
   if (verbose) edm::LogInfo("L1TStage2EMTF") << "L1TStage2EMTF: analyze..." << std::endl;
-
+  if (isData){
```
Comment (davidlange6) [inline DQM/L1TMonitor/src/L1TStage2EMTF.cc]: why is the isData needed? 
@EsmaeelEskandari

File: DQM/L1TMonitor/src/L1TdeStage2EMTF.cc [hunk 2]
```diff
@@ -104,24 +102,85 @@ void L1TdeStage2EMTF::analyze(const edm::Event& e, const edm::EventSetup& c) {
...
+
+  l1t::EMTFTrack closestData; 
+  l1t::EMTFTrack closestEmul;
+
+  //Find the pair of data/emulator tracks that minimizes d(phi)^2 + d(eta)^2 
+  for (std::vector<l1t::EMTFTrack>::const_iterator dataTrack = dataTrackCollection->begin(); dataTrack != dataTrackCollection->end(); ++dataTrack){
+    if (dataTrack->BX() > 1 || dataTrack->BX() < -1) continue;
+    for (std::vector<l1t::EMTFTrack>::const_iterator emulTrack = emulTrackCollection->begin(); emulTrack !=emulTrackCollection->end(); ++emulTrack){
+      if (emulTrack->BX() > 1 || emulTrack->BX() < -1) continue;
+      float dPhi = dataTrack->Phi_glob_rad() - emulTrack->Phi_glob_rad();
```
Comment (davidlange6) [inline DQM/L1TMonitor/src/L1TdeStage2EMTF.cc]: use DataFormats/Math/include/deltaPhi.h

File: DQM/L1TMonitor/src/L1TdeStage2EMTF.cc [hunk 3]
```diff
@@ -104,24 +102,85 @@ void L1TdeStage2EMTF::analyze(const edm::Event& e, const edm::EventSetup& c) {
...
+      if (dPhi > 3.14159) dPhi -= 3.14159;//ensure dPhi falls between -pi and +pi
+      if (dPhi < -3.14159) dPhi += 3.14159;
+      float dEta = dataTrack->Eta() - emulTrack->Eta();
+      if (dPhi*dPhi + dEta*dEta < mindR){
```
Comment (davidlange6) [inline DQM/L1TMonitor/src/L1TdeStage2EMTF.cc]: actually - everything can just be replaced by deltaR.h in the same package

File: DQM/L1TMonitorClient/src/L1TStage2EMTFDEClient.cc [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+  emtfTrackSectorIndexComp_=ibooker.book1D("emtfTrackSectorIndexComp","Data/Emul of Track Sector Index", 13, -6.5, 6.5);
+
+
+  for (int bin = 1; bin <= 16; ++bin) {
+    emtfTrackModeComp_->setBinLabel(bin, std::to_string(bin - 1), 1);
+    emtfTrackQualComp_->setBinLabel(bin, std::to_string(bin - 1), 1);
+  }
+} 
+void L1TStage2EMTFDEClient::fixZero(TH1F* histData, TH1F* histEmul){
+  for (Int_t i=0;i<histData->GetSize();i++){
```
Comment (davidlange6) [inline DQM/L1TMonitorClient/src/L1TStage2EMTFDEClient.cc]: please use native types not root ones.

File: DQM/L1TMonitorClient/src/L1TStage2EMTFDEClient.cc [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+    TH1F *hwEtaD = dataHist_->getTH1F();
+    TH1F *hwEtaE = emulHist_->getTH1F();
+
+    TH1F *hwEtaRatio = emtfMuonhwEtaComp_->getTH1F();
+    TH1F *hwEtaDif = (TH1F*) hwEtaRatio->Clone();
+    L1TStage2EMTFDEClient::fixZero(hwEtaD, hwEtaE); 
+
+    hwEtaDif->Add(hwEtaD, hwEtaE);
+
+    hwEtaRatio->Divide(hwEtaD, hwEtaDif, 2, 1);
```
Comment (davidlange6) [inline DQM/L1TMonitorClient/src/L1TStage2EMTFDEClient.cc]: so this effectively becomes

2*hwEtaD[i] / ( max(0.1+hwEtaD[i]) + max(0.1+hwEtaE) )

Right? Is the double 0.1 needed?

File: DQM/L1TMonitorClient/src/L1TStage2EMTFDEClient.cc [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+   //Mode
+  dataHist_ = igetter.get(input_dir_data_+"/"+"emtfTrackMode");
+  emulHist_ = igetter.get(input_dir_emul_+"/"+"emtfTrackMode");
+
+ if (dataHist_ && emulHist_){
+    TH1F *ModeD = dataHist_->getTH1F();
+    TH1F *ModeE = emulHist_->getTH1F();
+
+    TH1F *ModeRatio = emtfTrackModeComp_->getTH1F();
+    TH1F *ModeDif = (TH1F*) ModeRatio->Clone();
```
Comment (davidlange6) [inline DQM/L1TMonitorClient/src/L1TStage2EMTFDEClient.cc]: here and elsewhere, the Clone() needs to be deleted before going out of scope (or ownership given to something else)

**Comment by EsmaeelEskandari** [comment]
Dear @davidlange6, You are right but these codes works and their produced plots are validated and actually needed by experts. Changing, testing and validating them again needs some time that I haven't now. Would you please merge this PR? 
If it is really necessary to change the codes I can come back to them in 2017 January or February that we have no data taking.

**Comment by dmitrijus** [comment]
-1

Hi @EsmaeelEskandari,

it will not be merged until you address the comments.
That Clone() creates a memory leak too, L1T has been fighting those for quite a while now, so it is for everyone's best interest if you address the comments.

**Comment by EsmaeelEskandari** [comment]
@dmitrijus @davidlange6 @vanbesien @bortigno @abrinke1 @nhaubrich 
Why DQM core didn't checked the code so carfully like above at first time that I re-commited  
PR https://github.com/cms-l1t-offline/cmssw/pull/386 ( the ancestor of #16869 ) that was pushed to a wrong branch in July by @nhaubrich . 
From that time all EMTF DQM developers disappeared and despite of having small knowledge of EMTF, when Pier asked me to take care of that PR I accepted. I updated that PR more than 10 times to take into account DQM core and other expert people comments. Now It is working and it is validated and @davidlange6 is just taking about a more efficient code. I have no time any more to waste for this PR.
and really get tired of updating https://github.com/cms-l1t-offline/cmssw/pull/386 and if you don't want to merge it then someone else can take care of it or you can forget it.

**Comment by dmitrijus** [comment]
@EsmaeelEskandari 

DQM (me) did check the code. Unfortunately, several bugs slipped passed through.
This is the reason multiple people look at the code.

I know how hard it can be to write stable code, but it has to be done nevertheless. It is also a huge amount of work just to review your code, it isn't a 10 line change. By creating a new pull request every time you want fix something, isn't helping either. 

This changes aren't just for "more efficient code", this is a memory leak. This is something L1T DQM has been struggling with lately. Thank you, @davidlange6 for spotting it.
Related: https://github.com/cms-l1t-offline/cmssw/issues/423

It is way better to spot and fix such errors early on.   
Besides, if you don't have time to fix them now, I doubt you will have time to do it later.

Please make sure to join L1T DQM Workshop @ Dec. 15

**Comment by bortigno** [comment]
Hi all, 

Thanks for taking care of L1T code and for reviewing it. Having multiple people looking at our work is definitely helpful and we are all willing to fix problems and resolve bugs when found. 

We all had a very intense and successful year also thanks to your help and collaboration in running this PR ( actually #16737 ) online for most of the run without the need of integrating it in CMSSW: having eyes on target while running was the most important thing for us. Given that the run is over I think that for the final integration in CMSSW for next year @abrinke1  can take care of organising the fixes requested by this PR within the EMTF group.

Thank you

**Pull request closed without merging**